File: lib/models/user_progress.dart
import 'package:cloud_firestore/cloud_firestore.dart';

enum LearningLevel {
  level1, // Basic tile identification
  level2, // Understanding suits and sets
  level3, // Hand combinations
  level4, // Scoring fundamentals
  level5, // Advanced scoring
  level6, // Strategic gameplay
}

class LevelProgress {
  final LearningLevel level;
  final int completedLessons;
  final int totalLessons;
  final int quizScore;
  final bool isCompleted;
  final DateTime? completedAt;

  const LevelProgress({
    required this.level,
    required this.completedLessons,
    required this.totalLessons,
    required this.quizScore,
    this.isCompleted = false,
    this.completedAt,
  });

  double get progressPercentage =>
      totalLessons > 0 ? completedLessons / totalLessons : 0.0;

  Map<String, dynamic> toJson() => {
        'level': level.name,
        'completedLessons': completedLessons,
        'totalLessons': totalLessons,
        'quizScore': quizScore,
        'isCompleted': isCompleted,
        'completedAt': completedAt?.toIso8601String(),
      };

  factory LevelProgress.fromJson(Map<String, dynamic> json) => LevelProgress(
        level: LearningLevel.values.firstWhere(
          (e) => e.name == json['level'],
        ),
        completedLessons: json['completedLessons'] as int,
        totalLessons: json['totalLessons'] as int,
        quizScore: json['quizScore'] as int,
        isCompleted: json['isCompleted'] as bool? ?? false,
        completedAt: json['completedAt'] != null
            ? DateTime.parse(json['completedAt'] as String)
            : null,
      );
}

class UserProgress {
  final String userId;
  final String variant; // Selected mahjong variant
  final Map<LearningLevel, LevelProgress> levelProgress;
  final int totalTimeSpent; // in seconds
  final int gamesPlayed;
  final int gamesWon;
  final int quizzesCompleted;
  final List<String> achievements; // Achievement IDs
  final DateTime lastUpdated;
  final DateTime createdAt;

  const UserProgress({
    required this.userId,
    required this.variant,
    required this.levelProgress,
    this.totalTimeSpent = 0,
    this.gamesPlayed = 0,
    this.gamesWon = 0,
    this.quizzesCompleted = 0,
    this.achievements = const [],
    required this.lastUpdated,
    required this.createdAt,
  });

  double get overallProgress {
    if (levelProgress.isEmpty) return 0.0;
    final totalProgress = levelProgress.values
        .map((p) => p.progressPercentage)
        .fold(0.0, (a, b) => a + b);
    return totalProgress / levelProgress.length;
  }

  int get completedLevels =>
      levelProgress.values.where((p) => p.isCompleted).length;

  Map<String, dynamic> toJson() => {
        'userId': userId,
        'variant': variant,
        'levelProgress': levelProgress.map(
          (key, value) => MapEntry(key.name, value.toJson()),
        ),
        'totalTimeSpent': totalTimeSpent,
        'gamesPlayed': gamesPlayed,
        'gamesWon': gamesWon,
        'quizzesCompleted': quizzesCompleted,
        'achievements': achievements,
        'lastUpdated': Timestamp.fromDate(lastUpdated),
        'createdAt': Timestamp.fromDate(createdAt),
      };

  factory UserProgress.fromJson(Map<String, dynamic> json) {
    final levelProgressMap = <LearningLevel, LevelProgress>{};
    final levelData = json['levelProgress'] as Map<String, dynamic>?;
    if (levelData != null) {
      for (final entry in levelData.entries) {
        final level = LearningLevel.values.firstWhere(
          (e) => e.name == entry.key,
        );
        levelProgressMap[level] = LevelProgress.fromJson(entry.value);
      }
    }

    return UserProgress(
      userId: json['userId'] as String,
      variant: json['variant'] as String,
      levelProgress: levelProgressMap,
      totalTimeSpent: json['totalTimeSpent'] as int? ?? 0,
      gamesPlayed: json['gamesPlayed'] as int? ?? 0,
      gamesWon: json['gamesWon'] as int? ?? 0,
      quizzesCompleted: json['quizzesCompleted'] as int? ?? 0,
      achievements: List<String>.from(json['achievements'] as List? ?? []),
      lastUpdated: (json['lastUpdated'] as Timestamp).toDate(),
      createdAt: (json['createdAt'] as Timestamp).toDate(),
    );
  }

  UserProgress copyWith({
    String? userId,
    String? variant,
    Map<LearningLevel, LevelProgress>? levelProgress,
    int? totalTimeSpent,
    int? gamesPlayed,
    int? gamesWon,
    int? quizzesCompleted,
    List<String>? achievements,
    DateTime? lastUpdated,
    DateTime? createdAt,
  }) {
    return UserProgress(
      userId: userId ?? this.userId,
      variant: variant ?? this.variant,
      levelProgress: levelProgress ?? this.levelProgress,
      totalTimeSpent: totalTimeSpent ?? this.totalTimeSpent,
      gamesPlayed: gamesPlayed ?? this.gamesPlayed,
      gamesWon: gamesWon ?? this.gamesWon,
      quizzesCompleted: quizzesCompleted ?? this.quizzesCompleted,
      achievements: achievements ?? this.achievements,
      lastUpdated: lastUpdated ?? this.lastUpdated,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

File: src/models/UserProgress.ts
import {Timestamp} from '@react-native-firebase/firestore';

export enum LearningLevel {
  LEVEL1 = 'level1', // Basic tile identification
  LEVEL2 = 'level2', // Understanding suits and sets
  LEVEL3 = 'level3', // Hand combinations
  LEVEL4 = 'level4', // Scoring fundamentals
  LEVEL5 = 'level5', // Advanced scoring
  LEVEL6 = 'level6', // Strategic gameplay
}

export interface LevelProgress {
  level: LearningLevel;
  completedLessons: number;
  totalLessons: number;
  quizScore: number;
  isCompleted: boolean;
  completedAt?: Date;
}

export const calculateProgressPercentage = (progress: LevelProgress): number => {
  return progress.totalLessons > 0 
    ? progress.completedLessons / progress.totalLessons 
    : 0.0;
};

export interface UserProgress {
  userId: string;
  variant: string; // Selected mahjong variant
  levelProgress: Record<LearningLevel, LevelProgress>;
  totalTimeSpent: number; // in seconds
  gamesPlayed: number;
  gamesWon: number;
  quizzesCompleted: number;
  achievements: string[]; // Achievement IDs
  lastUpdated: Date;
  createdAt: Date;
}

export const calculateOverallProgress = (progress: UserProgress): number => {
  const levels = Object.values(progress.levelProgress);
  if (levels.length === 0) return 0.0;
  
  const totalProgress = levels.reduce((sum, p) => {
    return sum + calculateProgressPercentage(p);
  }, 0.0);
  
  return totalProgress / levels.length;
};

export const getCompletedLevels = (progress: UserProgress): number => {
  return Object.values(progress.levelProgress).filter(p => p.isCompleted).length;
};

export const userProgressToJson = (progress: UserProgress): Record<string, any> => {
  const levelProgressJson: Record<string, any> = {};
  Object.entries(progress.levelProgress).forEach(([key, value]) => {
    levelProgressJson[key] = {
      level: value.level,
      completedLessons: value.completedLessons,
      totalLessons: value.totalLessons,
      quizScore: value.quizScore,
      isCompleted: value.isCompleted,
      completedAt: value.completedAt?.toISOString(),
    };
  });

  return {
    userId: progress.userId,
    variant: progress.variant,
    levelProgress: levelProgressJson,
    totalTimeSpent: progress.totalTimeSpent,
    gamesPlayed: progress.gamesPlayed,
    gamesWon: progress.gamesWon,
    quizzesCompleted: progress.quizzesCompleted,
    achievements: progress.achievements,
    lastUpdated: Timestamp.fromDate(progress.lastUpdated),
    createdAt: Timestamp.fromDate(progress.createdAt),
  };
};

export const userProgressFromJson = (json: Record<string, any>): UserProgress => {
  const levelProgressMap: Record<LearningLevel, LevelProgress> = {} as Record<LearningLevel, LevelProgress>;
  const levelData = json.levelProgress as Record<string, any> | undefined;
  
  if (levelData) {
    Object.entries(levelData).forEach(([key, value]) => {
      const level = key as LearningLevel;
      levelProgressMap[level] = {
        level: value.level as LearningLevel,
        completedLessons: value.completedLessons as number,
        totalLessons: value.totalLessons as number,
        quizScore: value.quizScore as number,
        isCompleted: value.isCompleted as boolean ?? false,
        completedAt: value.completedAt ? new Date(value.completedAt as string) : undefined,
      };
    });
  }

  return {
    userId: json.userId as string,
    variant: json.variant as string,
    levelProgress: levelProgressMap,
    totalTimeSpent: (json.totalTimeSpent as number) ?? 0,
    gamesPlayed: (json.gamesPlayed as number) ?? 0,
    gamesWon: (json.gamesWon as number) ?? 0,
    quizzesCompleted: (json.quizzesCompleted as number) ?? 0,
    achievements: (json.achievements as string[]) ?? [],
    lastUpdated: (json.lastUpdated as Timestamp).toDate(),
    createdAt: (json.createdAt as Timestamp).toDate(),
  };
};

